source/common/upstream/upstream_impl.cc

```diff
@@ -1354,6 +1399,12 @@ StrictDnsClusterImpl::ResolveTarget::~ResolveTarget() {
   }
 }
 
+void StrictDnsClusterImpl::onHealthCheckHostRemoval(const HostSharedPtr& host) {
+  for (const auto& target : resolve_targets_) {
+    target->all_hosts_.erase(host->address()->asString());
```

Comment:
This opens up for some weird behavior in the case where multiple resolve targets resolve to the same ip: there might be multiple resolve targets that have a host for `1.1.1.1`, and since health checking is not shared between the two of them you'll see one fail HC before the other. At that point all hosts with ip `1.1.1.1` is removed, not just the one that failed health checks.

If host A and B both point to the same ip based on resolve target C and D, I think you could see the following sequence on the main thread:
Resolve target C removes A
Host A fails HC, A and B are removed from `all_hosts_` in C and D respectively
Resolve target D updates with the same hosts
Since B is no longer in all_hosts_, it looks like a new host so it replaced by a new host for `1.1.1.1`, causing connections to `1.1.1.1` to be drained, health check value reset etc.

I think the quick fix is to do something like 
``` 
auto itr = target.all_hosts_.find(host->address()->asString());
if (itr != target.all_hosts_.end() && itr->second == host) {
  target.all_hosts_.erase(itr);
}
```
, which unfortunately doesn't make this any less gross. 

The longer term fix to this problem that I've been toying with involves splitting up the `Host` into the load assignment piece (where does it fit within the LB structures) and the endpoint pieces (the ip:port you're making connections to and health check) so that hc/connection state can be shared between lb objects (this relates to #4280). This would eliminate the need for `all_hosts_` completely, as each lb object could just reuse the per endpoint state from the shared endpoint object. 

Unfortunately this turns out to be quite a large undertaking due to how widely used `Host`/`HostDescription` is throughout the code base, so a simpler medium term solution would be great.

Reply:
Thanks for catching this. I've thought about it a bit, and I'm wondering if I should just ignore DNS for now. In practice with a reasonable polling rate I don't think it's a big deal. If I implement this only in EDS I think I can do the virtual calls in a way that aren't ugly and we can avoid the all_hosts mess in DNS. WDYT?

Reply:
Yeah I think omitting this for DNS is fine